src/data/transform.ts

```diff
@@ -0,0 +1,555 @@
+      excalidrawElements.add(element as ExcalidrawElement);
+
+      return;
+    }
+    const elementWithId = { ...element, id: elementId };
+
+    if (
+      (elementWithId.type === "rectangle" ||
+        elementWithId.type === "ellipse" ||
+        elementWithId.type === "diamond" ||
+        elementWithId.type === "arrow") &&
+      elementWithId?.label?.text
+    ) {
+      let [container, text] = bindTextToContainer(
+        elementWithId as
+          | ValidContainer
+          | ({
+              type: "arrow";
+            } & ValidLinearElement),
+        elementWithId?.label,
+      );
+      excalidrawElements.add(container);
+      excalidrawElements.add(text);
+
+      if (container.type === "arrow") {
+        const originalStart =
+          elementWithId.type === "arrow" ? elementWithId?.start : undefined;
+        const originalEnd =
+          elementWithId.type === "arrow" ? elementWithId?.end : undefined;
+        const { linearElement, startBoundElement, endBoundElement } =
+          bindLinearElementToElement({
+            ...container,
+            start: originalStart,
+            end: originalEnd,
+          });
+        container = linearElement;
+        excalidrawElements.add(linearElement);
+        excalidrawElements.add(startBoundElement);
+        excalidrawElements.add(endBoundElement);
+      }
+    } else {
+      let excalidrawElement;
+      if (elementWithId.type === "text") {
+        const fontFamily = elementWithId?.fontFamily || DEFAULT_FONT_FAMILY;
+        const fontSize = elementWithId?.fontSize || DEFAULT_FONT_SIZE;
+        const lineHeight =
+          elementWithId?.lineHeight || getDefaultLineHeight(fontFamily);
+        const text = elementWithId.text ?? "";
+        const normalizedText = normalizeText(text);
+        const metrics = measureText(
+          normalizedText,
+          getFontString({ fontFamily, fontSize }),
+          lineHeight,
+        );
+        excalidrawElement = {
+          width: metrics.width,
+          height: metrics.height,
+          fontFamily,
+          fontSize,
+          ...elementWithId,
+        };
+
+        excalidrawElements.add(excalidrawElement as ExcalidrawTextElement);
+      } else if (elementWithId.type === "arrow") {
+        const { linearElement, startBoundElement, endBoundElement } =
+          bindLinearElementToElement(elementWithId);
+        excalidrawElements.add(linearElement);
+        excalidrawElements.add(startBoundElement);
+        excalidrawElements.add(endBoundElement);
+      } else if (elementWithId.type === "line") {
+        const width = elementWithId.width || DEFAULT_LINEAR_ELEMENT_PROPS.width;
+        const height =
+          elementWithId.height || DEFAULT_LINEAR_ELEMENT_PROPS.height;
+        const lineElement = newLinearElement({
+          width,
+          height,
+          points: [
+            [0, 0],
+            [width, height],
+          ],
+          ...elementWithId,
```

Comment:
should be spread on top, so that whatever the user defines doesn't override the computer properties

Reply:
here the intention is to override the properties and give more preference to the user instead

Reply:
if that's the case and we don't care about the user potentially overriding the properties we've properly re-calculated, then we should align all the other cases where we spread the original element?

One other thing that _may not_ happen now, but could happen later: we're not stripping the custom properties when we spread the objects.

Reply:
> if that's the case and we don't care about the user potentially overriding the properties we've properly re-calculated, then we should align all the other cases where we spread the original element?

Can you elaborate on this ?

> One other thing that may not happen now, but could happen later: we're not stripping the custom properties when we spread the objects.

Where I am calling `newElement`, `newLinearElement` etc APIs there it strips off so we need not worry however there are two places where I am not doing thats regular text and shapes (rectangle, ellipse, diamond) and we are not supporting any custom attributes in these cases so it doesn't matter but I can strip of custom attributes if passed mistakenly eg passing `start` for `rectangle`

Reply:
> Can you elaborate on this ?

https://github.com/excalidraw/excalidraw/blob/06dd6de0aff19f1728d55fee6312a3d5d6e6d03d/src/data/transform.ts#L545-L550

There are other cases where when we calculate e.g. the wrapped text, you always prefer that calculated text. And I agree. That's why I think we should also prefer calculated `width`/`height` and other properties and not let them be overridden by the user.

> [...] we are not supporting any custom attributes in these cases so it doesn't matter but I can strip of custom attributes if passed mistakenly eg passing `start` for `rectangle`

Yep, hence why I said it may not happen now, but later (when we introduce custom props to generic or other elements). 

If we only care about the editor, `restore()` will take care of it. But if we care about the return type of the transformer, we may want to pass the generic elements through `newElement()` as well.